published/external/xdpapi_helper.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include "xdpapi.h"
+
+#define TEST_TIMEOUT_ASYNC_MS 1000
+#define POLL_INTERVAL_MS 10
+
+static
+VOID
+XdpDetach(
+    _In_ VOID *ClientContext
+    )
+{
+    UNREFERENCED_PARAMETER(ClientContext);
+}
+
+NTSTATUS
+InitializeXdpApi(
+    _Out_ XDP_API_CLIENT *ApiContext,
+    _Out_ const XDP_API_PROVIDER_DISPATCH **ProviderDispatch,
+    _Out_ const XDP_API_PROVIDER_BINDING_CONTEXT **ProviderBindingContext,
+    _In_ const XDP_API_CLIENT_DISPATCH *ClientDispatch,
+    _In_ UINT32 Version
+    )
+{
+    NTSTATUS Status;
+    KEVENT Event;
+    INT32 TimeoutMs = TEST_TIMEOUT_ASYNC_MS;
+
+    Status = XdpLoadApi(Version, NULL, NULL, &XdpDetach, ClientDispatch, ApiContext);
+    if (!NT_SUCCESS(Status)) {
+        return Status;
+    }
+
+    KeInitializeEvent(&Event, NotificationEvent, FALSE);
+
+    do {
+        LARGE_INTEGER Timeout100Ns;
+
+        Status =
+            XdpOpenApi(
+                ApiContext,
+                ProviderDispatch,
+                ProviderBindingContext);
+        if (NT_SUCCESS(Status)) {
+            break;
+        }
+
+        Timeout100Ns.QuadPart = -1 * Int32x32To64(POLL_INTERVAL_MS, 10000);
+        KeResetEvent(&Event);
```

Comment:
I still don't understand the point of this event - where does it get set?

Reply:
As mentioned, sleep before retrying.
@nigriMSFT should know details

Reply:
Looks like this code is implementing a sleep by intentionally timing out in KeWaitForSingleObject